### ISSUE TITLE
Improve battle engine efficiency with large number of units

### ISSUE THREAD

**Issue by lanedirt** [lanedirt/OGameX#406]
Issue #403 noticed a problem when a battle occurs with high number of units (100K+). After investigation it seemed the error was caused by an out-of-memory exception. The PHP max memory limit was set to 128MB. PR #404 fixes this by increasing the max memory limit to 1GB instead, as well as a small optimization to the `BattleUnit` class.

Still, with the above fixes a battle with 700K+ units can take upwards of 1 minute to process. This is too slow.

Also if we look at certain unit counts in the official game such as this Reddit post shows, the game engine should be able to support (tens of) milions of units in a single battle.
https://www.reddit.com/r/OGame/comments/1gata35/i_finaly_did_it_1m_pts/

Related issues:
- #503 

## Todo:

- [ ] Improve BattleEngine logic located in `app/GameMissions/BattleEngine/BattleEngine.php` to make it more efficient for larger unit numbers.

## Note:
A feature test has already been added which tests this usecase for a fleet with a large amount of units. This test currently runs for about 60 seconds (on a local dev docker instance). This test can be used as a starting point for testing performance of the (to be improved) BattleEngine. See this file:

`tests/Feature/FleetDispatch/FleetDispatchAttackTest.php` --> test method `testLargeScaleAttackWithDebrisField()`.

**Comment by jackbayliss** [comment]
I'm curious how it works in the original, does it just load immediately? 

There's potential to offload large battles to a job  which can process in the background. Noticed you've already created a Rust PR though which seems that it'll be more performant

**Comment by lanedirt** [comment]
I'm not sure how the original takes on this problem as I do not have any current accounts with large amount of units. But from my experience large battles could take a few seconds before they appeared, but not much more than that.

The current OGameX PHP implementation already struggles with a few million units taking up to 60 seconds to process. Checking some hall of fame battle reports in the original game forums show battles can have tens of millions of units. So offloading it to a more performant compiled language seems the best way to go for now.

My aim is to make the new Rust implementation backwards compatible and interchangeable with the existing PHP version. So if any new better ideas come up we should be able to swap it out fairly easy.

I haven't worked with Rust before so figuring it out as I go, but in theory it should be a lot more performant.

**Comment by jackbayliss** [comment]
Sounds good to me in that case. Thanks for the reply :)

**Comment by setube** [comment]
Bro, you might want to try asynchronous processing or message queues.

**Comment by lanedirt** [comment]
> Bro, you might want to try asynchronous processing or message queues.

Yes agreed, implementing a message queue of some sorts is needed to improve the processing of missions including attack missions (battles), I created an issue for this some time ago, see #363 for more details.

**Comment by lanedirt** [comment]
I have made quite a bit of progress on the Rust battle engine in PR #517, all unit tests are now passing with the new Rust battle engine implementation.

The good news is that the new Rust Battle engine version already shows a dramatic improvement. The `Tests\Feature\FleetDispatch\FleetDispatchAttackTest` test on my Macbook Pro M2 takes:

- PHP battle engine: 53.24sec
- Rust battle engine: 0.78sec

```
www@1a887b4602b1:/var/www$ php artisan test tests/Feature/FleetDispatch/FleetDispatchAttackTest.php --filter="withdebris"

   PASS  Tests\Feature\FleetDispatch\FleetDispatchAttackTest
  ✓ large scale attack with debris field                                                                                               53.20s  

  Tests:    1 passed (17 assertions)
  Duration: 53.24s

www@1a887b4602b1:/var/www$ php artisan test tests/Feature/FleetDispatch/FleetDispatchAttackTest.php --filter="withdebris"

   PASS  Tests\Feature\FleetDispatch\FleetDispatchAttackTest
  ✓ large scale attack with debris field                                                                                                0.74s  

  Tests:    1 passed (17 assertions)
  Duration: 0.78s
```

So that's already a 67x performance improvement which is dramatically quicker. I'll need to run some more tests in terms of memory usage and do some general cleanup and refactoring, but it's looking good so far.

**Comment by dep1engenharia** [comment]
Great @lanedirt!! 
That's a amazing performance boost. Rust is a very powerfull programming language for this kind of tasks.
